<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Rvtests by zhanxw</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Rvtests</h1>
        <h2>Rare variant test software for next generation sequencing data</h2>
        <a href="https://github.com/zhanxw/rvtests" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          

<p><strong>Table of Contents</strong></p>

<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#download">Download</a></li>
<li>
<a href="#quick-tutorial">Quick tutorial</a>

<ul>
<li><a href="#single-variant-tests">Single variant tests</a></li>
<li><a href="#groupwise-tests">Groupwise tests</a></li>
<li><a href="#related-individual-tests">Related individual tests</a></li>
<li>
<a href="#meta-analysis-tests">Meta-analysis tests</a>

<ul>
<li><a href="#dominant-models-and-recessive-models">Dominant models and recessive models</a></li>
</ul>
</li>
</ul>
</li>
<li>
<a href="#input-files">Input files</a>

<ul>
<li><a href="#genotype-file-vcf">Genotype file (VCF)</a></li>
<li><a href="#phenotype-file">Phenotype file</a></li>
<li><a href="#covariate-file">Covariate file</a></li>
<li><a href="#trait-transformation">Trait transformation</a></li>
</ul>
</li>
<li>
<a href="#models">Models</a>

<ul>
<li><a href="#single-variant-tests">Single variant tests</a></li>
<li><a href="#burden-tests">Burden tests</a></li>
<li><a href="#variable-threshold-models">Variable threshold models</a></li>
<li><a href="#kernel-models">Kernel models</a></li>
<li><a href="#meta-analysis-models">Meta-analysis models</a></li>
<li><a href="#utility-models">Utility models</a></li>
</ul>
</li>
<li>
<a href="#association-test-options">Association test options</a>

<ul>
<li><a href="#sample-inclusionexclusion">Sample inclusion/exclusion</a></li>
<li>
<a href="#variant-site-filters">Variant site filters</a>

<ul>
<li><a href="#annotation">Annotation</a></li>
</ul>
</li>
<li><a href="#genotype-filters">Genotype filters</a></li>
<li><a href="#handle-missing-genotypes-and-phenotypes">Handle missing genotypes and phenotypes</a></li>
<li><a href="#specify-groups-eg-burden-unit">Specify groups (e.g burden unit)</a></li>
</ul>
</li>
<li><a href="#sex-chromosome-analysis">Sex chromosome analysis</a></li>
<li><a href="#kinship-generation">Kinship generation</a></li>
<li>
<a href="#resources">Resources</a>

<ul>
<li><a href="#ucsc-refflat-genes">UCSC RefFlat Genes</a></li>
<li><a href="#gencode-genes">Gencode Genes</a></li>
</ul>
</li>
<li><a href="#frequently-asked-questions-faq">Frequently Asked Questions (FAQ)</a></li>
<li><a href="#feedbackcontact">Feedback/Contact</a></li>
</ul>



<p><a href="https://travis-ci.org/zhanxw/rvtests"><img src="https://travis-ci.org/zhanxw/rvtests.png?branch=master" alt="Build Status"></a></p>

<p>(Updated: November 2015)</p>

<h1>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h1>

<p>Rvtests, which stands for Rare Variant tests, is a flexible software package for genetic association studies. It is designed to support unrelated individual or related (family-based) individuals. Both quantitative trait and binary trait are supported. It includes a variety of association tests (e.g. single variant score test, burden test, variable threshold test, SKAT test, fast linear mixed model score test). It takes <a href="http://www.1000genomes.com/">VCF</a> format as genotype input file and takes PLINK format phenotype file and covariate file. From our practice, it is capable to analyze 8,000 related individuals using less than 400 Mb memory. </p>

<h1>
<a id="download" class="anchor" href="#download" aria-hidden="true"><span class="octicon octicon-link"></span></a>Download</h1>

<p>Source files can be downloaded from <a href="https://github.com/zhanxw/rvtests/archive/master.zip">github</a> or <a href="https://github.com/zhanxw/rvtests">github page</a>.
Executable binary files (for Linux 64bit) can be downloaded from <a href="https://github.com/zhanxw/rvtests/releases">here</a>.</p>

<h1>
<a id="quick-tutorial" class="anchor" href="#quick-tutorial" aria-hidden="true"><span class="octicon octicon-link"></span></a>Quick tutorial</h1>

<p>Here is a quick example of how to use <em>rvtests</em> software in typical use cases.</p>

<h2>
<a id="single-variant-tests" class="anchor" href="#single-variant-tests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Single variant tests</h2>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --out output --single wald,score
</code></pre>

<p>This specifies single variant Wald and score test for association
tests for every variant in the <code>input.vcf</code> file. The 6th column of the phenotype file, <code>phenotype.ped</code>, which is in PLINK format, is used. Rvtests will automatically check whether the phenotype is binary trait or quantitative trait.
For binary trait, the recommended way of coding is to code controls as 1, cases as 2, missing phenotypes as -9 or 0.</p>

<p>For other types of association tests, you can refer to <a href="#models">Models</a>.</p>

<h2>
<a id="groupwise-tests" class="anchor" href="#groupwise-tests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Groupwise tests</h2>

<p>Groupwise tests includes three major kinds of tests.</p>

<ul>
<li>Burden tests: group variants, which are usually less than 1% or 5% rare variants, for association tests. The category includes: CMC test, Zeggini test, Madsen-Browning test, CMAT test, and rare-cover test.</li>
<li>Variable threshold tests: group variants under different frequency thresholds.</li>
<li>Kernel methods: suitable to tests rare variants having different directions of effects. These includes SKAT test and KBAC test. </li>
</ul>

<p>All above tests requires to group variants into a unit. The simplest case is to use gene as grouping unit. For different grouping method, see <a href="#Grouping">Grouping</a>. </p>

<p>To perform rare variant tests by gene, you need to use <code>--geneFile</code> to specify the gene range in a refFlat format. We provided different gene definitions in the <a href="#Resources">Resources</a> section. You can use <code>--gene</code> to specify which gene(s) to test. For example, specify <code>--gene CFH,ARMS2</code> will perform association tests on CFH and ARMS2 genes. If there is no providing <code>--gene</code> option, all genes will be tests.</p>

<p>The following command line demonstrate how to use CMC method, variable threshold method(proposed by Price) and kernel based method (SKAT by Shawn Lee and KBAC by
Dajiang Liu) to test every gene listed in <em>refFlat_hg19.txt.gz</em>.</p>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --out output --geneFile refFlat_hg19.txt.gz --burden cmc --vt price --kernel skat,kbac
</code></pre>

<h2>
<a id="related-individual-tests" class="anchor" href="#related-individual-tests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Related individual tests</h2>

<p>To test related individuals, you will need to first create a kinship matrix:</p>

<pre><code>vcf2kinship --inVcf input.vcf --bn --out output
</code></pre>

<p>The option <code>--bn</code> means calculating empirical kinship using Balding-Nicols method. You can specify <code>--ibs</code> to obtain IBS kinship or use <code>--pedigree input.ped</code> to calculate kinship from known pedigree information.</p>

<p>Then you can use linear mixed model based association tests such as Fast-LMM score test, Fast-LMM LRT test and Grammar-gamma tests. An exemplar command is shown: </p>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --out output --kinship output.kinship --single famScore,famLRT,famGrammarGamma
</code></pre>

<h2>
<a id="meta-analysis-tests" class="anchor" href="#meta-analysis-tests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Meta-analysis tests</h2>

<p>The meta-analysis models outputs association test results and genotype covariance matrix. These statistics can be used in rare variant association analysis.
We provide single variant score test and generate genotype covariance matrix. 
You can use command:</p>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --meta score,cov --out output
</code></pre>

<p>In a more realistic scenario, you may want to adjust for covariates and want to inverse normalized residuals obtained in null model (<a href="http://www.nature.com/ng/journal/vaop/ncurrent/full/ng.2852.html">link</a> to our methodology paper), then this command will work:</p>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --covar example.covar --covar-name age,bmi --inverseNormal --useResidualAsPhenotype  --meta score,cov --out output
</code></pre>

<p>Here the <code>--covar</code> specify a covariate file, and <code>--covar-name</code> specify which covariates can used in the analysis. Covariate file format can be found <a href="#Covariate%20file">here</a>. <code>--inverseNormal --useResidualAsPhenotype</code> specifies trait transformation method. That means first fit a regression model of the phenotype on covariates (intercept automatically added), then the residuals are inverse normalized. Trait transformation details can be found <a href="#Trait%20transformation">here</a>.</p>

<p>We support both unrelated individuals and related individuals (e.g. family data). You need to append <code>--kinship input.kinship</code> to the command line:</p>

<pre><code>rvtests --inVcf input.vcf --pheno phenotype.ped --meta score,cov --out output --kinship input.kinship
</code></pre>

<p>The file <code>input.kinship</code> is calculated by <code>vcf2kinship</code> program, and usage to this program is described in <a href="#related-individual-tests">Related individual tests</a>.</p>

<h3>
<a id="dominant-models-and-recessive-models" class="anchor" href="#dominant-models-and-recessive-models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dominant models and recessive models</h3>

<p>Dominant and recessive disease models are supported by appending "dominant" and/or "recessive" after "--meta" option. For example, use "--meta dominant,recessive" will 
generate two sets of files. For dominant model, they are "prefix.MetaDominant.assoc" and "prefix.MetaDominantCov.assoc.gz"; for recessive model,
they are "prefix.MetaRecessive.assoc" and "prefix.MetaRecessiveCov.assoc.gz". Internally, in dominant models, genotypes 0/1/2 are coded as 0/1/1; in recessive models, genotypes 0/1/2 are 
coded as 0/0/1. Missing genotypes will be imputed to the mean.</p>

<h1>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input files</h1>

<h2>
<a id="genotype-file-vcf" class="anchor" href="#genotype-file-vcf" aria-hidden="true"><span class="octicon octicon-link"></span></a>Genotype file (VCF)</h2>

<p>Rvtests supports VCF (Variant Call Format) files. Files in both plain text format or gzipped format are supported. To use group-based rare variant tests, indexed the VCF files using <a href="http://samtools.sourceforge.net/tabix.shtml">tabix</a> are required. </p>

<p>Here are the commands to convert plain text format to bgzipped VCF format:</p>

<pre><code>(grep ^"#" $your_old_vcf; grep -v ^"#" $your_old_vcf | sed 's:^chr::ig' | sort -k1,1n -k2,2n) | bgzip -c &gt; $your_vcf_file 
tabix -f -p vcf $your_vcf_file
</code></pre>

<p>The above commands will (1) remove the <code>chr</code> prefix from chromosome names; (2) sort VCF files by chromosome first, then by chromosomal positions; (3) compress using bgzip; (4) create tabix index.</p>

<p>Rvtests support genotype dosages. Use <code>--dosage DosageTag</code> to specify the dosage tag. For example, if VCF format field is "GT:EC" and individual genotype fields is "0/0:0.02", you can use <code>--dosage EC</code>, and rvtests will use the dosage 0.02 in the regression models.</p>

<h2>
<a id="phenotype-file" class="anchor" href="#phenotype-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>Phenotype file</h2>

<p>You can use <code>--mpheno $phenoypeColumnNumber</code> or <code>--pheno-name</code> to specify a given phenotype.</p>

<p>An example phenotype file, (<code>example.pheno</code>), has the following format: </p>

<pre><code>fid iid fatid matid sex y1 y2 y3 y4
P1 P1 0 0 0 1.7642934435605 -0.733862638327895 -0.980843608339726 2
P2 P2 0 0 0 0.457111744989746 0.623297281416372 -2.24266162284447 1
P3 P3 0 0 0 0.566689682543218 1.44136462889459 -1.6490100777089 1
P4 P4 0 0 0 0.350528353203767 -1.79533911725537 -1.11916876241804 1
P5 P5 0 0 1 2.72675074738545 -1.05487747371158 -0.33586430010589 2
</code></pre>

<p>Phenotype file is specified by the option <code>--pheno example.pheno</code> . The default phenotype column header is “<code>y1</code>”. If you want to use alternative columns as phenotype for association analysis (e.g the column with header y2), you may specify the phenotype by column or by name using either</p>

<ul>
<li>--mpheno 2 </li>
<li>--pheno-name y2</li>
</ul>

<p><strong>NOTE:</strong> to use “<code>--pheno-name</code>”, the  header line must starts with “<code>fid iid</code>” as PLINK requires.</p>

<p>In phenotype file, missing values can be denoted by NA or any non-numeric values. Individuals with missing phenotypes will be automatically dropped from subsequent association analysis. For each missing phenotype value, a warning will be generated and recorded in the log file.</p>

<p>When the phenotype values are only 0, 1 and 2, rvtests will automatically treat it as binary traits. However, if you want to treat it as continuous trait, please use "<code>--qtl</code>" option.</p>

<h2>
<a id="covariate-file" class="anchor" href="#covariate-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>Covariate file</h2>

<p>You can use <code>--covar</code> and <code>--covar-name</code> to specify covariates that will be used for single variant association analysis. This is an optional parameter. If you do not have covariate in the data, this option can be ignored. </p>

<p>The covariate file, (e.g. <code>example.covar</code>) has a similar format as the phenotype file:</p>

<pre><code>fid iid fatid matid sex y1 y2 y3 y4
P1 P1 0 0 0 1.911 -1.465 -0.817 1
P2 P2 0 0 0 2.146 -2.451 -0.178 2
P3 P3 0 0 0 1.086 -1.194 -0.899 1
P4 P4 0 0 0 0.704 -1.052 -0.237 1
P5 P5 0 0 1 2.512 -3.085 -2.579 1
</code></pre>

<p>The covariate file is specified by the <code>--covar</code> option (e.g. <code>--covar example.covar</code>). To specify covariates that will be used in the association analysis, the option <code>--covar-name</code> can be used. For example, when age, bmi and 3 PCs are used for association analysis, the following option can be specified for the rvtests program, i.e. 
<code>--covar example.covar --covar-name age,bmi,pc1,pc2,pc3</code>.</p>

<p>Note: Missing data in the covariate file can be labeled by any non-numeric value (e.g. NA). They will be automatically imputed to the mean value in the data file. </p>

<h2>
<a id="trait-transformation" class="anchor" href="#trait-transformation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Trait transformation</h2>

<p>In this meta-analysis, we use inverse normal transformed residuals in the association analysis, which is achieved by using a combination of <code>--inverseNormal</code>  and <code>--useResidualAsPhenotype</code>. Specifically, we first fit the null model by regressing phenotype on covariates. The residuals are then inverse normal transformed (see Appendix A more detailed formula for transformation). Transformed residuals will be used to obtain score statistics. </p>

<p>In meta analysis, an exemplar command for using rvtests looks like the following:</p>

<pre><code>./rvtest --inVcf $vcf --pheno $example.pheno --covar example.covar --covar-name age,bmi --inverseNormal --useResidualAsPhenotype  --meta score,cov --out $output_prefix  
</code></pre>

<h1>
<a id="models" class="anchor" href="#models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Models</h1>

<p>Rvtests support various association models.</p>

<h2>
<a id="single-variant-tests-1" class="anchor" href="#single-variant-tests-1" aria-hidden="true"><span class="octicon octicon-link"></span></a>Single variant tests</h2>

<table>
<thead>
<tr>
<th align="left">Single variant</th>
<th align="center">Model(#)</th>
<th align="center">Traits(##)</th>
<th align="center">Covariates</th>
<th align="center">Related / unrelated</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Score test</td>
<td align="center">score</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Only null model is used to performed the test</td>
</tr>
<tr>
<td align="left">Wald  test</td>
<td align="center">wald</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Only fit alternative model, and effect size will be estimated</td>
</tr>
<tr>
<td align="left">Exact test</td>
<td align="center">exact</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Fisher's test (allelic test)</td>
</tr>
<tr>
<td align="left">Dominant Exact test</td>
<td align="center">dominantExact</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Fisher's test (dominant codings)</td>
</tr>
<tr>
<td align="left">Fam LRT</td>
<td align="center">famLRT</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">Fast-LMM model</td>
</tr>
<tr>
<td align="left">Fam Score</td>
<td align="center">famScore</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">Fast-LMM model style likelihood ratio test</td>
</tr>
<tr>
<td align="left">Grammar-gamma</td>
<td align="center">famGrammarGamma</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">Grammar-gamma method</td>
</tr>
<tr>
<td align="left">Firth regression</td>
<td align="center">firth</td>
<td align="center">B</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Logistic regression with Firth correction by David Firth, discussed by Clement Ma.</td>
</tr>
</tbody>
</table>

<p>(#) Model columns list the recognized names in rvtests. For example, use <code>--single score</code> will apply score test.</p>

<p>(##) In trait column, B or Q stand for binary or quantitative trait, respectively.</p>

<h2>
<a id="burden-tests" class="anchor" href="#burden-tests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Burden tests</h2>

<table>
<thead>
<tr>
<th align="left">Burden tests</th>
<th align="center">Model(#)</th>
<th align="center">Traits(##)</th>
<th align="center">Covariates</th>
<th align="center">Related / unrelated</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">CMC</td>
<td align="center">cmc</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Collapsing and combine rare variants by Bingshan Li.</td>
</tr>
<tr>
<td align="left">Zeggini</td>
<td align="center">zeggini</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Aggregate counts of rare variants by Morris Zeggini.</td>
</tr>
<tr>
<td align="left">Madsen-Browning</td>
<td align="center">mb</td>
<td align="center">B</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Up-weight rare variant using inverse frequency from controls by Madsen.</td>
</tr>
<tr>
<td align="left">Fp</td>
<td align="center">fp</td>
<td align="center">B</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Up-weight rare variant using inverse frequency from controls by Danyu Lin.</td>
</tr>
<tr>
<td align="left">Exact CMC</td>
<td align="center">exactCMC</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Collapsing and combine rare variants, then perform Fisher's exact test.</td>
</tr>
<tr>
<td align="left">CMC Wald</td>
<td align="center">cmcWald</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Collapsing and combine rare variants, then perform Wald test.</td>
</tr>
<tr>
<td align="left">RareCover</td>
<td align="center">rarecover</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Find optimal grouping unit for rare variant tests by Thomas Hoffman.</td>
</tr>
<tr>
<td align="left">CMAT</td>
<td align="center">cmat</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Test non-coding variants by Matt Zawistowski.</td>
</tr>
<tr>
<td align="left">FamCMC</td>
<td align="center">famcmc</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R</td>
<td align="left">Collapsing and combine rare variants extended to related samples.</td>
</tr>
<tr>
<td align="left">FamZeggini</td>
<td align="center">famzeggini</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R</td>
<td align="left">Aggregate counts of rare variants extended to related samples.</td>
</tr>
</tbody>
</table>

<p>(#) Model columns list the recognized names in rvtests. For example, use <code>--burden cmc</code> will apply CMC test.</p>

<p>(##) In trait column, B or Q stand for binary or quantitative trait, respectively.</p>

<h2>
<a id="variable-threshold-models" class="anchor" href="#variable-threshold-models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Variable threshold models</h2>

<table>
<thead>
<tr>
<th align="left">Single variant</th>
<th align="center">Model(#)</th>
<th align="center">Traits(##)</th>
<th align="center">Covariates</th>
<th align="center">Related / unrelated</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Variable threshold model by permutation</td>
<td align="center">price</td>
<td align="center">B, Q</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Every rare-variant frequency cutoffs are tests by Alkes Price.</td>
</tr>
<tr>
<td align="left">Variable threshold model by analytic form</td>
<td align="center">analytic</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Every rare-variant frequency cutoffs are tests by Danyu Lin.</td>
</tr>
<tr>
<td align="left">Variable threshold model by analytic form</td>
<td align="center">famAnalytic</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R</td>
<td align="left">Every rare-variant frequency cutoffs are tests by Dajiang Liu.</td>
</tr>
</tbody>
</table>

<p>(#) Model columns list the recognized names in rvtests. For example, use <code>--vt price</code> will variable threshold test.
NOTE: our implementatino of Price's test diffs from the original method descrbied in Price's publcation. We test every minor allele frequency cutoff (instead of reference allele counts) and this is a two-sided test (instead of one-sided test).</p>

<p>(##) In trait column, B or Q stand for binary or quantitative trait, respectively.</p>

<h2>
<a id="kernel-models" class="anchor" href="#kernel-models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Kernel models</h2>

<table>
<thead>
<tr>
<th align="left">Kernel</th>
<th align="center">Model(#)</th>
<th align="center">Traits(##)</th>
<th align="center">Covariates</th>
<th align="center">Related / unrelated</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">SKAT</td>
<td align="center">skat</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Sequencing kernel association test by Shawn Lee.</td>
</tr>
<tr>
<td align="left">SKATO</td>
<td align="center">skato</td>
<td align="center">B, Q</td>
<td align="center">Y</td>
<td align="center">U</td>
<td align="left">Optimal sequencing kernel association test (SKAT-O) by Shawn Lee. (###)</td>
</tr>
<tr>
<td align="left">KBAC</td>
<td align="center">kbac</td>
<td align="center">B</td>
<td align="center">N</td>
<td align="center">U</td>
<td align="left">Kernel-based adaptive clustering model by Dajiang Liu.</td>
</tr>
<tr>
<td align="left">FamSKAT</td>
<td align="center">famSkat</td>
<td align="center">Q</td>
<td align="center">Y</td>
<td align="center">R</td>
<td align="left">Sequencing kernel association test extended to related individuals by Han Chen.</td>
</tr>
</tbody>
</table>

<p>(#) Model columns list the recognized names in rvtests. For example, use <code>--kernel skat</code> will apply SKAT test.
To further customize SKAT test, you can use <em>--kernel skat[nPerm=100:alpha=0.001:beta1=1:beta2=20]</em> to specify permutation counts, type-1 error, 
beta distribution parameters for up-weighting rare variants. Rvtests will output a message showing: </p>

<pre><code>[INFO]  SKAT test significance will be evaluated using 10000 permutations at alpha = 0.001 (beta1 = 1.00, beta2 = 20.00)
</code></pre>

<p>(##) In trait column, B or Q stand for binary or quantitative trait, respectively.</p>

<p>(###) SKAT-O implementation may have slightly different results compared with outputs from SKAT R package. That's probably due to numerical stability.</p>

<h2>
<a id="meta-analysis-models" class="anchor" href="#meta-analysis-models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Meta-analysis models</h2>

<table>
<thead>
<tr>
<th align="left">Type</th>
<th align="center">Model(#)</th>
<th align="center">Traits(##)</th>
<th align="center">Covariates</th>
<th align="center">Related / unrelated</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Score test</td>
<td align="center">score</td>
<td align="center">B,Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">standard score tests</td>
</tr>
<tr>
<td align="left">Dominant model</td>
<td align="center">dominant</td>
<td align="center">B,Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">score tests and covariance matrix under dominant disease model</td>
</tr>
<tr>
<td align="left">Recessive model</td>
<td align="center">recessive</td>
<td align="center">B,Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">score tests and covariance matrix under recessive disease model</td>
</tr>
<tr>
<td align="left">Covariance</td>
<td align="center">cov</td>
<td align="center">B,Q</td>
<td align="center">Y</td>
<td align="center">R, U</td>
<td align="left">covariance matrix</td>
</tr>
</tbody>
</table>

<p>(#) Model columns list the recognized names in rvtests. For example, use <code>--meta score,cov</code> will generate score statistics and covariance matrix for meta-analysis.</p>

<p>(##) In trait column, B or Q stand for binary or quantitative trait, respectively.</p>

<p>The above models are suitable to generate summary statistics which can be later meta-analyzed (see <a href="http://www.nature.com/ng/journal/v46/n2/abs/ng.2852.html">Dajiang Liu (2014) Nature Genetics</a>).
Rvtests implemented the above methods and the results can be further analyzed by RareMetals (<a href="http://genome.sph.umich.edu/wiki/RareMETALS">link</a>) for quantitative trait and RareMetals2 (<a href="http://genome.sph.umich.edu/wiki/RareMETALS2">link</a>).
It also worth to mention that our group offers another toolset for meta analysis (<a href="http://genome.sph.umich.edu/wiki/Rare-Metal">link</a>).</p>

<p><strong>Explanation of outputs</strong></p>

<p>The meta-analysis results come as two files: summary score statistics files (prefix.MetaScore.assoc.gz) and covariance files (prefix.MetaCov.assoc.gz).</p>

<p>In summary score statistics files, you will obtain these columns:</p>

<ul>
<li>N_INFORMATIVE: Number of samples that are analyzed for association. </li>
<li>AF: allele frequency. For related individuals, we use BLUE estimator. For case-control study, we list overall frequency (adjusted by relatedness if possible), case frequency and control frequency separated by a colon.</li>
<li>INFORMATIVE_ALT_AC: The number of alternative alleles in the analyzed samples.</li>
<li>CALL_RATE: The fraction of non-missing alleles. For case-control study, we calculate call rate for all samples, case samples and control samples separated by a colon.</li>
<li>HWE_PVALUE: Hardy-Weinberg equilibrium. For related individuals, this statistic can be inflated. For case-control study, we calculate HWE pvalues for all samples, case samples and controls samples separated by a colon.</li>
<li>N_REF/N_HET/N_ALT: Number of samples carrying homozygous reference/heterozygous/homozygous alternative alleles. For case-control study, we calculate these three statistics for all samples, case samples and controls samples separated by a colon.</li>
<li>U_STAT, SQRT_V_STAT: U and V statistics are score statistics and their covariance matrix. Details can be found in <a href="http://www.nature.com/ng/journal/v46/n2/abs/ng.2852.html">Dajiang Liu (2014) Nature Genetics</a>.</li>
<li>ALT_EFFSIZE: for continuous outcome, this is the estimated effect size; for binary outcome, this is the estimated log odds-ratio. We apply a new correction method when binary trait associations for
related individuals are analyzed in standard linear mixed models. The log odds ratio is approximately correct for related individual analysis as well.</li>
</ul>

<p>The covariance files stores covariances in sliding windows. You will obtain these columns:</p>

<ul>
<li>CHROM: the chromosome name</li>
<li>START_POS/END_POS: within one sliding window, the first/last variant chromosomal position.</li>
<li>NUM_MARKER: number of markers in one sliding window.</li>
<li>MARKER_POS: all chromosomal positions in the sliding window.</li>
<li>COV: covariances between markers. When the sliding window has markers G.1, G.2, ..., G.N, this column will store "covariances" of (G.1, G.1), (G.1, G.2), ..., (G.1, G.N);
for binary outcomes, this column uses colons as separators and additionally stores covariances between N genotypes and K covariates (G.1, C.1), (G.1, C.2), ..., (G.1, C.K), as well as
covariances between K covariates (C.1, C.1), (C.1, C.2), ... (C.1, C.K), (C.2, C.2), ..., (C.2, C.K), ..., (C.K, C.K).</li>
</ul>

<h2>
<a id="utility-models" class="anchor" href="#utility-models" aria-hidden="true"><span class="octicon octicon-link"></span></a>Utility models</h2>

<p>Rvtests has an convinient option <code>--outputRaw</code>. When specifing this, rvtests can output genotypes, phenotype, covariates (if any) and collapsed genotype to tabular files. These files can be imported into other software (e.g. R) for further analyses.</p>

<h1>
<a id="association-test-options" class="anchor" href="#association-test-options" aria-hidden="true"><span class="octicon octicon-link"></span></a>Association test options</h1>

<h2>
<a id="sample-inclusionexclusion" class="anchor" href="#sample-inclusionexclusion" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample inclusion/exclusion</h2>

<p>Rvtests can flexibly specify which sample(s) to include or exclude:</p>

<pre><code>       --peopleIncludeID : give IDs of people that will be included in study
     --peopleIncludeFile : from given file, set IDs of people that will be included in study
       --peopleExcludeID : give IDs of people that will be included in study
     --peopleExcludeFile : from given file, set IDs of people that will be included in study
</code></pre>

<p><code>--peopleIncludeID</code> and <code>--peopleExcludeID</code> are used to include/exclude samples from command line. 
For example, specify <code>--peopleIncludeID A,B,C</code> will include A, B and C sample from the VCF files if they exists.
<code>--peopleIncludeID</code> and <code>--peopleExcludeID</code> followed by a file name will include or exclude the IDs in the file.
So to include sample A, B and C, you can provide a file, <code>people.txt</code>, looks like:</p>

<pre><code>A
B
C
</code></pre>

<p>Then use <code>--peopleIncludeFile people.txt</code> to include them in the analysis.</p>

<h2>
<a id="variant-site-filters" class="anchor" href="#variant-site-filters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Variant site filters</h2>

<p>It is common that different frequency cutoffs are applied in rare-variant analysis.
Therefore, rvtests specify frequency cutoffs.</p>

<p>Frequency Cutoff</p>

<pre><code>             --freqUpper : Specify upper minor allele frequency bound to be included in analysis
             --freqLower : Specify lower minor allele frequency bound to be included in analysis
</code></pre>

<p>If you specify <code>--freqLower 0.01 --freqUpper 0.05</code>, only the variants with minor allele frequency between 0.01 and 0.05 (boundary inclusive) will be analyzed.</p>

<p>Similar to sample inclusion/exclusion options, you can specify a range of variants to be included by 
specifying <code>--rangeList</code> option. For example <code>--rangeList 1:100-200</code> will include the chromosome 1 position 100bp to 200bp region.
Alternatively, use a separate file, <code>range.txt</code>, and <code>--rangeFile range.txt</code> to specify association tests range.</p>

<pre><code>             --rangeList : Specify some ranges to use, please use chr:begin-end format.
             --rangeFile : Specify the file containing ranges, please use chr:begin-end format.
              --siteFile : Specify the file containing sites to include, please use "chr pos" format.
</code></pre>

<p>It is supported to filter variant site by site depth, minor allele count or annotation (annotated VCF file is needed).</p>

<pre><code>          --siteDepthMin : Specify minimum depth(inclusive) to be included in analysis
          --siteDepthMax : Specify maximum depth(inclusive) to be included in analysis
            --siteMACMin : Specify minimum Minor Allele Count(inclusive) to be included in analysis
              --annoType : Specify annotation type that is followed by ANNO= in the VCF INFO field, regular expression is allowed
</code></pre>

<p><em>NOTE</em>: <code>--annoType</code> can filter variants based on regular expression.
For example, <code>--annoType Nonsynonymous</code> will only analyze non-synonymous variants where they have <code>ANNO=Nonsynonymous</code> in the INFO field.
To extract more than one annotation types, use <code>--annoType 'Start_Gain|Stop_Loss|Start_Loss|Essential_Splice_Site|Stop_Gain|Normal_Splice_Site|Synonymous|Nonsynonymous</code> will extract LOF (loss of function) mutations.
To generate an annotated VCF file, please read on the next section <a href="#Annotation">Annotation</a>. </p>

<h3>
<a id="annotation" class="anchor" href="#annotation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Annotation</h3>

<p>We define a VCF file with annotating information as an annotated VCF here.
The annotation step can be done using <a href="https://github.com/zhanxw/anno">ANNO</a>, a fast and accurate annotation software.
At minimal, there are two steps to annotate a VCF file:</p>

<ol>
<li>
<p>Install ANNO and its resources files</p>

<p>git clone <a href="https://github.com/zhanxw/anno.git">https://github.com/zhanxw/anno.git</a>
cd anno; make
cd resources; ./download.sh</p>
</li>
<li>
<p>Run the following script:</p>

<p>anno -i input.vcf -o output.vcf.gz -r anno/resources/hs37d5.fa -g anno/resources/refFlat_hg19.txt.gz -p anno/priority.txt -c anno/codon.txt --indexOutput</p>
</li>
</ol>

<p>You will then obtain an annotated VCF file <code>output.vcf.gz</code> and its tabix index <code>output.vcf.gz.tbi</code>.</p>

<h2>
<a id="genotype-filters" class="anchor" href="#genotype-filters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Genotype filters</h2>

<p>Genotype with low depth or low quality can be filtered out by these options:</p>

<pre><code>          --indvDepthMin : Specify minimum depth(inclusive) of a sample to be included in analysis
          --indvDepthMax : Specify maximum depth(inclusive) of a sample to be included in analysis
           --indvQualMin : Specify minimum depth(inclusive) of a sample to be included in analysis
</code></pre>

<p>When genotypes are filtered, they are marked as missing genotypes. 
Consequently, samples with missing genotype may or may not be included in the analysis.
That means samples with genotypes may be dropped (<code>--impute drop</code>) 
or may still be included (<code>--impute mean</code> or <code>--impute hwe</code>). 
By default, genotypes are imputed to its means.
See next section about how you like to handle missing genotypes.</p>

<h2>
<a id="handle-missing-genotypes-and-phenotypes" class="anchor" href="#handle-missing-genotypes-and-phenotypes" aria-hidden="true"><span class="octicon octicon-link"></span></a>Handle missing genotypes and phenotypes</h2>

<p>When genotypes are missing (e.g. genotype = "./.") or genotypes are filtered out, 
there are three options to handle them: (1) impute to its mean(default option); (2) impute by HWE equilibrium; (3) remove from the model.
Use <code>--impute [mean|hwe|drop]</code> to specify which option to use.</p>

<p>When quantitative phenotypes are missing, for example, some samples have genotype files, but not phenotypes, 
rvtests can impute missing phenotype to its mean. </p>

<p><em>NOTE:</em> Do not use <code>--imputePheno</code> for binary trait.</p>

<p>In summary, the following two options can be used:</p>

<pre><code>           --impute : Specify either of mean, hwe, and drop
      --imputePheno : Impute phenotype to mean by those have genotypes but no
                      phenotypes
</code></pre>

<h2>
<a id="specify-groups-eg-burden-unit" class="anchor" href="#specify-groups-eg-burden-unit" aria-hidden="true"><span class="octicon octicon-link"></span></a>Specify groups (e.g burden unit)</h2>

<p>Rare variants association tests are usually performed in groups of variants. 
The natural grouping unit is gene. Rvtests can read gene definition file in <code>refFlat</code> format,
and perform association for each gene. Use <code>--geneFile</code> option to specify the gene file name.
For example, <code>--geneFile refFlat_hg19.txt.gz</code> will use <code>refFlat_hg19.txt.gz</code> as gene definition file,
and then perform association tests for every gene. Use <code>--gene</code> to specify a subset of genes to test.
For example, <code>--gene CFH</code> will only test CFH gene.</p>

<p>Alternative grouping unit can be specified as <em>set</em>. 
These <em>sets</em> are treated similar to gene.
You can thus use <code>--setFile</code> to define sets (similar to <code>--geneFile</code> option), 
and use <code>--set</code> to define a specific set (similar to <code>--gene</code> option). 
Additionally, use <code>--setList</code> can specify a set to test from command line.</p>

<p>The format of a set file is: (1) set names; (2) ranges (e.g. chrom:begin-end);
For example, you have a set file, <code>example.set</code>, like this:</p>

<pre><code>set1 1:100-200,1:250-300
set2 2:500-600
</code></pre>

<p>You can specify <code>--setFile example.set --set set2</code> to group variants 
within chromosome 2, position 500 to 600bp. 
If you want to test a particular region, for example, chromosome 2, position 500 to 550bp,
but do not want to make another file, you can use <code>--setList 2:500-600</code>.</p>

<p>In summary, options related to <em>Grouping Unit</em> are listed below: </p>

<pre><code>         --geneFile : specify a gene file (for burden tests)
             --gene : specify which genes to test
          --setList : specify a list to test (for burden tests)
          --setFile : specify a list file (for burden tests, first two columns:
                      setName chr:beg-end)
              --set : specify which set to test (1st column)
</code></pre>

<h1>
<a id="sex-chromosome-analysis" class="anchor" href="#sex-chromosome-analysis" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sex chromosome analysis</h1>

<p>Rvtests support X chromosome analysis. In human X chromosome, there is PAR (pseudoautosomal region) and non-PAR region.
For males, there are two X allele in PAR region and one allele in non-PAR region.
While the PAR region is treated in the same way as autosomes, rvtests treat non-PAR region differently.
Below we will describe the details about how rvtests handles non-PAR region.</p>

<p><em>Prepare data</em>. According to VCF standard, male genotype needs to coded as 0 or 1. For compatibility, rvtests also support 0/0 or 1/1 coding. 
In VCF files, male genotypes can be written as "0", "1", "0|0", "0/0", "1|1", "1/1". All other genotypes will be treated as missing.</p>

<p><em>Genotype in the regression model</em>. For consistence, male genotypes are converted to 0 or 2.</p>

<p><em>MetaScore results</em>. If <code>--meta score</code> is specified, the output file <code>prefix.MetaScore.assoc.gz</code> includes PAR-region and non-PAR region analysis. 
However, in the non-PAR region, the difference is that Hardy-Weinberg P-value and homeozygous-reference/heterzygous/homozygous-alternative sample sizes are calculated using female samples only.</p>

<p><em>Related individuals</em>. Just append <code>--xHemi</code> to the <code>vcf2kinship</code> (more details in <a href="#kinship-generation">Kinship generation</a>) and <code>rvtest</code> command lines. Rvtests can recognize non-PAR region kinship file and use it in the analysis.</p>

<p><em>PAR region</em>. PAR region is defined as two regions X:60001-2699520 and X:154931044-155260560. Use <code>--xLabel</code> can specify which chromosome has PAR region (default: 23 or X)
and use <code>--xParRegion</code> to specify PAR region (default: hg19, meaning '60001-2699520,154931044-155260560' in the UCSC build hg19, specify "hg18" will use PAR region definition in the UCSC build hg18, or specify "hg38" will use UCSC build 38).</p>

<h1>
<a id="kinship-generation" class="anchor" href="#kinship-generation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Kinship generation</h1>

<p>Analysis of related individual usually requires estimation of kinship. You can a separate tool, <code>vcf2kinship</code>.
<code>vcf2kinship</code> is usually included in rvtests binary distribution or can be built from software source codes.</p>

<p><code>vcf2kinship</code> can calculate pedigree kinship using a pedigree input file (PED format, see <a href="#phenotype-file">Phenotype file</a>, use option <code>--ped</code>).
The output file name is specified by <code>--prefix</code> option. If you use <code>--prefix output</code> then the output files will include <code>output.kinship</code>.</p>

<p>It can also calculate empirical kinship using genotype input file (VCF format, see <a href="#genotype-file-vcf">Genotype file (VCF)</a>, use option <code>--inVcf</code>).
For empirical kinship, you also need to specify the kinship model, either Balding-Nicols model (use option <code>--bn</code>) or Identity-by-state model (use option <code>--ibs</code>).</p>

<p>In sex chromosome analysis, it is often required to generate kinship on X chromosome regions, then you need to specify <code>--xHemi</code>. If your input VCF file has different X chromosome label (e.g. chromosome name is '23' instead of 'X'), you can use <code>--xLabel 23</code>.</p>

<p>If principal component decomposition (PCA) results are needed, you can use option <code>--pca</code>. Then output files with suffix '.pca' include PCA results.</p>

<p>When dealing with large input files, it is often preferred to use multiple CPU to speed up calculation using the option <code>--thread N</code> in which N is the number of CPU.</p>

<p>For example, to generate pedigree-based kinship (<code>--ped</code>) on both autosomal region and X chromosome (<code>--xHemi</code>) region, the command line is:</p>

<pre><code>vcf2kinship --ped input.ped --xHemi --out output
</code></pre>

<p>To generate empirical kinship (<code>--inVcf</code>) on both autosomal region and X chromosome (<code>--xHemi</code>) region using Balding-Nicols model, the command line is:</p>

<pre><code>vcf2kinship --inVcf input.vcf.gz --ped input.ped --bn --xHemi --out output
</code></pre>

<p>NOTE: you need to provide a pedigree file (PED) in the above case, as <code>vcf2kinship</code> need the sex information of samples.</p>

<h1>
<a id="resources" class="anchor" href="#resources" aria-hidden="true"><span class="octicon octicon-link"></span></a>Resources</h1>

<h2>
<a id="ucsc-refflat-genes" class="anchor" href="#ucsc-refflat-genes" aria-hidden="true"><span class="octicon octicon-link"></span></a>UCSC RefFlat Genes</h2>

<p>refFlat_hg19.txt.gz</p>

<p>UCSC gene definition file in the refFlat format (<a href="http://genome.ucsc.edu/goldenpath/gbdDescriptionsOld.html#RefFlat">Details</a>) in NCBI genome build 37 .</p>

<p>File link: <a href="http://qbrc.swmed.edu/zhanxw/seqminer/data/refFlat_hg19.txt.gz">http://qbrc.swmed.edu/zhanxw/seqminer/data/refFlat_hg19.txt.gz</a></p>

<h2>
<a id="gencode-genes" class="anchor" href="#gencode-genes" aria-hidden="true"><span class="octicon octicon-link"></span></a>Gencode Genes</h2>

<p>refFlat.gencode.v19.gz</p>

<p>Gencode gene definition version 19 in the refFlat format (<a href="http://www.gencodegenes.org/releases/19.html">Details</a>). We have also previous versions of gene files and can provide upon request.</p>

<p>File link: <a href="http://qbrc.swmed.edu/zhanxw/seqminer/data/refFlat.gencode.v19.gz">http://qbrc.swmed.edu/zhanxw/seqminer/data/refFlat.gencode.v19.gz</a></p>

<h1>
<a id="frequently-asked-questions-faq" class="anchor" href="#frequently-asked-questions-faq" aria-hidden="true"><span class="octicon octicon-link"></span></a>Frequently Asked Questions (FAQ)</h1>

<ul>
<li>Does rvtests support binary traits of related-individuals?</li>
</ul>

<p>Yes and no. Proper analyses of related-individual are supported in meta-analysis model as described in section (#meta-analysis-models).
In many other association models, supporting binary traits for related individuals is complex and we have not found good solutions.</p>

<ul>
<li>Can you provide a list of command line options?</li>
</ul>

<p>Rvtests have build-in help that can be found by executing <code>rvtest --help</code>.
We also put all available options in this <a href="https://github.com/zhanxw/rvtests/wiki/Command-Line-Options">link</a>.</p>

<ul>
<li>Can you provide standard error (SE) or confidence interval (CI) for the estimated Beta in the score model?</li>
</ul>

<p>In the output of MetaScore model (--meta score), the standard error is the inverse of SQRT_V_STAT.
For example, if SQRT_V_STAT = 2, that means the standard error of estimated beta is 1/2 = 0.5.</p>

<ul>
<li>Why the INFORMATIVE_ALT_AC, N_REF and N_ALT columns have zero counts for certain chromosome X regions in meta-analysis models?</li>
</ul>

<p>These counts are calculated from female individuals. If your study only has male samples, rvtests cannot report these counts. Because if a male carries a non-reference allele, we cannot conclude that this is heterozygous (0/1) site or homozygous alternatives (1/1) site.</p>

<ul>
<li>Why P-values can be -1?</li>
</ul>

<p>If rvtests fails to fit using a certain model, it cannot calculate P-value reliably. Rvtests will print P-value as -1 instead of any number between 0 and 1 to indicate that an error occurred.
However, this should rarely happen. Please contact us if you have further questions.</p>

<ul>
<li>Why SKAT Q-values reported by rvtests are different from the SKAT R package?</li>
</ul>

<p>We strictly follow the notations in the SKAT publication (<a href="http://www.hsph.harvard.edu/skat/">Wu et al. (2011) AJHG</a>). However, in SKAT R package, its implementation is slightly different.
For example, in quantitative trait anlaysis, Q is divided by (2 * \hat{sigma2}) in the R package, but not in rvtests.
Although Q values can be different, the P-values from the two software packges should match (only in rare cases, numerical accrucy may cause minor differences).</p>

<ul>
<li>How rvtests handle multi-allelic variants?</li>
</ul>

<p>In rvtests, we focus on bi-allelic, and thus treat multi-allelic variants as bi-allelic variants. Any genotype that includes other than reference allele and the first alternative allele will be treated as missing.
For example, genotyep '0/2' will be treated as missing.</p>

<h1>
<a id="feedbackcontact" class="anchor" href="#feedbackcontact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Feedback/Contact</h1>

<p>Questions and requests can be sent to
Xiaowei Zhan (<a href="mailto:zhanxw@umich.edu" title="mailto:zhanxw@umich.edu">zhanxw@umich.edu</a>)
or
Dajiang Liu (<a href="mailto:dajiang.liu@outlook.com" title="mailto:dajiang.liu@outlook.com">dajiang.liu@outlook.com</a>)
or
Goncalo Abecasis (<a href="mailto:goncalo@umich.edu" title="mailto:goncalo@umich.edu">goncalo@umich.edu</a>)</p>

<p>Rvtests is a collaborative effort from Xiaowei Zhan, Youna Hu, Bingshan Li, Dajiang Liu and Goncalo Abecasis.</p>

<p>We wawnt to thank rvtests users and especially those who have provided valuable feedbacks. These users include: Xueling Sim, Scott Verize, Shuang Feng, Kevin Lu, Ruth Loos, Tessel Galesloot, Valerie Turcot, Stefan Gustafsson, Corbin Quick, Adam Locke, and the GIANT consoritum and the GLGC consortium.</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/zhanxw/rvtests/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/zhanxw/rvtests/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/zhanxw/rvtests"></a> is maintained by <a href="https://github.com/zhanxw">zhanxw</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-21871925-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
